GVRf/Framework/src/org/gearvrf/skinning/GVRSkeleton.java

```diff
@@ -0,0 +1,215 @@
+
+public class GVRSkeleton implements PrettyPrint {
+    private static final String TAG = GVRSkeleton.class.getSimpleName();
+
+    protected GVRContext gvrContext;
+    protected GVRSceneObject sceneRoot;
+    protected GVRSceneObject animatedNode;
+    protected GVRSceneObject[] skeleton;
+    protected Matrix4f[] boneOffsetInverse;
+    protected Map<GVRSceneObject, List<GVRBone>> boneMap;
+
+    /**
+     * Constructs the skeleton for a list of {@link GVRSceneObject}.
+     *
+     * @param sceneRoot The scene root.
+     * @param animatedNodes The list of scene objects to animate.
+     */
+    public GVRSkeleton(GVRContext gvrContext, GVRSceneObject sceneRoot, GVRSceneObject animatedNode) {
+        this.gvrContext = gvrContext;
+        this.sceneRoot = sceneRoot;
+        this.animatedNode = animatedNode;
+
+        Set<GVRSceneObject> skeletalNodes = new HashSet<GVRSceneObject>();
+        boneMap = new HashMap<GVRSceneObject, List<GVRBone>>();
+
+        prepareSkeleton(animatedNode, skeletalNodes);
+
+        skeleton = new GVRSceneObject[skeletalNodes.size()];
+        boneOffsetInverse = new Matrix4f[skeletalNodes.size()];
+
+        Cell<Integer> boneIdx = new Cell<Integer>(0);
+        sortBones(skeletalNodes, sceneRoot, -1, boneIdx);
+        if (boneIdx.get() != skeletalNodes.size()) {
+            Log.w(TAG, "Skeleton cannot be fully constructed (number of bones = %d, bone array size = %d)",
+                    skeletalNodes.size(), boneIdx.get());
+        } else {
+            Log.v(TAG, "Skeleton is fully constructed with %d bones", boneIdx.get());
+        }
+    }
+
+    /**
+     * Update bone transform at each animation step.
+     */
+    public void animate(GVRKeyFrameAnimation animation, float animationTime) {
+        // Visit nodes in topological order
+        for (int nodeId = 0; nodeId < skeleton.length; ++nodeId) {
+            GVRSceneObject node = skeleton[nodeId];
+
+            // This node is not for a bone, so no need to compute its transform.
+            // Bone transform is relative to the holding scene object's frame.
+            if (boneMap.get(node) == null)
+                continue;
+
+            Matrix4f animationTransform;
+            if (animation != null) {
+                // update nodeTransform
+                animationTransform = animation.getTransform(node.getName(), animationTime);
+                if (animationTransform == null) {
+                    animationTransform = new Matrix4f();
+                }
+            } else {
+                // Debug: blank animation (z-axis rotation)
+                animationTransform = new Matrix4f();
+                animationTransform.rotation((float)(animationTime / 180 * Math.PI),
+                                            new Vector3f(1f, 0, 0));
+            }
+
+            Matrix4f boneTransform = new Matrix4f();
+
+            // Apply animation to bone
+            boneTransform.mul(boneOffsetInverse[nodeId]);
```

Comment:
i tried to get rid of math in java but it is hard. i will write a version using joml -- there is more convincing use-cases of joml when animation is added.

then, i will see if there is any way to replace it. if you know any good math library designed for openGL, please point out. I did good keywords like "java glm" and this is a top choice i can find.